pkg/controller/shared_state.go

```diff
@@ -1135,20 +1135,37 @@ func (c *SharedState) cleanUpAzVolumeAttachmentByNode(ctx context.Context, azDri
 func (c *SharedState) cleanUpAzVolumeAttachments(ctx context.Context, attachments []azdiskv1beta2.AzVolumeAttachment, cleanUp attachmentCleanUpMode, caller operationRequester) error {
 	var err error
 
+	var markDetachRequest = func(attachment *azdiskv1beta2.AzVolumeAttachment, patchRequired *bool) {
+		attachment.Status.Annotations = azureutils.AddToMap(attachment.Status.Annotations, consts.VolumeDetachRequestAnnotation, string(caller))
+		*patchRequired = true
+	}
+
+	var markCleanUp = func(attachment *azdiskv1beta2.AzVolumeAttachment, patchRequired *bool) {
+		attachment.Status.Annotations = azureutils.AddToMap(attachment.Status.Annotations, consts.CleanUpAnnotation, string(caller))
+		*patchRequired = true
+	}
+
 	for _, attachment := range attachments {
 		var patchRequired bool
 		patched := attachment.DeepCopy()
 
-		// if caller is azdrivernode, don't append cleanup annotation
-		if (caller != azdrivernode && !metav1.HasAnnotation(patched.ObjectMeta, consts.CleanUpAnnotation)) ||
-			// replica attachments should always be detached regardless of the cleanup mode
-			((cleanUp == cleanUpAttachment || patched.Spec.RequestedRole == azdiskv1beta2.ReplicaRole) && !metav1.HasAnnotation(patched.ObjectMeta, consts.VolumeDetachRequestAnnotation)) {
-			patchRequired = true
-			if caller != azdrivernode {
-				patched.Status.Annotations = azureutils.AddToMap(patched.Status.Annotations, consts.CleanUpAnnotation, string(caller))
+		if attachment.Spec.RequestedRole == azdiskv1beta2.PrimaryRole {
+			if cleanUp == cleanUpAttachment && !volumeDetachRequested(patched) {
+				markDetachRequest(patched, &patchRequired)
+			} else if deleteRequested, _ := objectDeletionRequested(&attachment); !deleteRequested {
+				// if primary azvolumeattachments are being cleaned up for driver uninstall, issue a DELETE call and continue
```

Comment:
np: add an explanation in the comment that the DELTE only deletes the CRI and doesn't issue detach of the disk first.